### PR TITLE
BracesPositionFixer - fix performance issue for massive files

### LINKED ISSUE #8828: BracesPositionFixer very slow for big migration file
### Problem description

`BracesPositionFixer` became super slow in https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/releases/tag/v3.79.0

With 3.78.1 it's like 6 seconds and with 3.79.0 over 20 minutes

### Minimal reproducer CLI command

bin/php-cs-fixer fix --rules='{"braces_position":true}' migrations/Version123.php

### Minimal reproducer code snippet

https://github.com/simoheinonen/php-cs-fixer-reproducer/blob/master/migrations/Version123.php

```php
<?php

namespace Application\Migrations;

use Doctrine\DBAL\Schema\Schema;
use Doctrine\Migrations\AbstractMigration;

class Version123 extends AbstractMigration
{
    public function up(Schema $schema): void
    {
        $this->addSql("
/* some sql */ {$foo} /* some sql */;
/* some sql */ {$foo} /* some sql */;
/* some sql */ {$foo} /* some sql */;

/* ~50k of these */

/* some sql */ {$foo} /* some sql */;
/* some sql */ {$foo} /* some sql */;
/* some sql */ {$foo} /* some sql */;
        ");
    }
}
```

### Fixer version

3.79.0

### PHP version

8.3 and 8.4

### How do you run Fixer?

Composer package

### Contribution Checks

- [x] I have verified if this problem was already reported
- [x] I am familiar with ["Feature or bug?"](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/blob/master/feature-or-bug.rst)

### PR DESCRIPTION
**Pull request by keradus**
closes #8828

### CODE REVIEW

File: tests/Fixer/Basic/BracesPositionFixerTest.php [hunk 1]
```diff
@@ -951,6 +963,31 @@ public static function provideFix84Cases(): iterable
...
+                    }
+                }
+                PHP,
+            <<<'PHP'
+                <?php class C
+                {
+                    private int $i = 5       {
+                        set(int $i)
+                        {
+                            $this->i = $i;
```
Comment (kubawerlos) [inline tests/Fixer/Basic/BracesPositionFixerTest.php]:
```suggestion
                    private mixed $wat = [-7, "a" . 'b', null, array(), 1/3] {
                        set(int $wat)
                        {
                            $this->wat = $wat;
                        }
                    }
                }
                PHP,
            <<<'PHP'
                <?php class C
                {
                    private mixed $wat = [-7, "a" . 'b', null, array(), 1/3]        {
                        set(int $wat)
                        {
                            $this->wat = $wat;
```
Let's emphasise that the tokens between `=` and `{` could be of a lot of different kinds and we cannot have a reasonable list of them.

File: tests/Fixtures/Integration/misc/issue_8828.test-out.php
Comment (kubawerlos) [inline tests/Fixtures/Integration/misc/issue_8828.test-out.php:1]: What about instead of this file, which is 1.7MB, we add a test case like below?
```php
        yield 'string with a lot of variables in braces' => [
            '<?php
                if (true) {
                    $foo = "'            . str_repeat(' text {$variable} text ', 50_000) . '";
                }',
        ];
```
Especially that the name of the test case is a spoiler for what is coming next: the current fix works for 👆🏼 , but not for 👇🏼 .
```php
        yield 'string with a lot of variables' => [
            '<?php
                if (true) {
                    $foo = "'            . str_repeat(' text $variable text ', 50_000) . '";
                }',
        ];
```
Which would need another over 1MB if we go in the direction of explicit files.

File: src/Fixer/Basic/BracesPositionFixer.php [hunk 2]
```diff
@@ -251,7 +251,7 @@ protected function applyFix(\SplFileInfo $file, Tokens $tokens): void
 
                 $positionOption = 'control_structures_opening_brace';
             } elseif ($token->isGivenKind(T_VARIABLE)) {
-                $openBraceIndex = $tokens->getNextTokenOfKind($index, ['{', ';', [CT::T_PROPERTY_HOOK_BRACE_OPEN], [T_CLOSE_TAG]]);
+                $openBraceIndex = $tokens->getNextTokenOfKind($index, ['{', ';', [CT::T_CURLY_CLOSE], [CT::T_PROPERTY_HOOK_BRACE_OPEN], [T_CLOSE_TAG]]);
```
Comment (kubawerlos) [inline src/Fixer/Basic/BracesPositionFixer.php]:
```suggestion
                $nextToken = $tokens->getNextMeaningfulToken($index);
                if (!$tokens[$nextToken]->equalsAny(['=', [CT::T_PROPERTY_HOOK_BRACE_OPEN]])) {
                    continue;
                }

                $openBraceIndex = $tokens->getNextTokenOfKind($x, ['{', ';', [CT::T_PROPERTY_HOOK_BRACE_OPEN], [T_CLOSE_TAG]]);
```
I was thinking about something like this:
- first - quick "deal breaker" for having a property hook after the variable
- second - proper check if the property hook is there

The `$x` could be either `$index` (for simplicity) or `$nextToken - 1` (for microoptimization).

**Review by kubawerlos — CHANGES_REQUESTED** [review]

File: tests/Fixtures/Integration/misc/issue_8828.test-out.php
Comment (keradus) [reply tests/Fixtures/Integration/misc/issue_8828.test-out.php:1]: tbh, i'm ok with both.

File: src/Fixer/Basic/BracesPositionFixer.php (hunk 2 above)
Comment (keradus) [reply src/Fixer/Basic/BracesPositionFixer.php]: not sure if I grasp your idea here, want to sent a PR with it ?

File: tests/Fixer/Basic/BracesPositionFixerTest.php (hunk 1 above)
Comment (keradus) [reply tests/Fixer/Basic/BracesPositionFixerTest.php]: i have a feeling that it can be waaaaay more bullshit inside. including the new edge cases.
i'm happy to move forward and unblock never-ending (20mins) fixing, by improving the covered range of complexity - while we can still face more in the future and fix separately.

File: tests/Fixtures/Integration/misc/issue_8828.test-out.php
Comment (keradus) [reply tests/Fixtures/Integration/misc/issue_8828.test-out.php:1]: i think to move back to integration test.
CI is failing [when code-coverage or mutation](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/actions/runs/16214828423/job/45781774207?pr=8830#step:12:1549), and integration test is outside of them by default

Comment (gharlan) [reply tests/Fixtures/Integration/misc/issue_8828.test-out.php:1]: I guess the same problem exists for concatenation of many variables..

```php
        yield 'string with concatenation of a lot of variables' => [
            '<?php
                if (true) {
                    $foo = "'            . str_repeat(' text ".$variable." text ', 50_000) . '";
                }',
        ];
```

File: src/Fixer/Basic/BracesPositionFixer.php (hunk 2 above)
Comment (gharlan) [reply src/Fixer/Basic/BracesPositionFixer.php]: If the variable is not followed by `=` or `T_PROPERTY_HOOK_BRACE_OPEN`, it cannot be a property with hooks.
In the given test cases it would be a fast fail for all variables inside the string.
Only for the first `$foo =  ...` variable it would still traverse through all tokens in the long statement to the statement end (`;`).

File: tests/Fixtures/Integration/misc/issue_8828.test-out.php
Comment (keradus) [reply tests/Fixtures/Integration/misc/issue_8828.test-out.php:1]: > i think to move back to integration test. CI is failing [when code-coverage or mutation](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/actions/runs/16214828423/job/45781774207?pr=8830#step:12:1549), and integration test is outside of them by default

done

**Comment by keradus** [comment]
I'm thinking to merge and release before my hols, so production setup is bit better.
further improvements can be raised as separated PRs
